File: apps/api/app/api/routes/ai_assistant.py
"""
Module 10 -- AI Assistant (LLD "Module: AI Assistant"): `POST /ai/assistant/
message`, `POST /ai/assistant/actions/{id}/confirm`, `GET /ai/assistant/
conversations/{id}`.

Permission gating per docs/ux/07-role-permission-matrix.md: `ai_assistant:
use` gates sending messages and reading conversations; `ai_assistant:
confirm_write` additionally gates confirming (not cancelling -- discarding
a proposal is strictly less capable than confirming one, so it's covered
by the same `ai_assistant:use` every role that can chat already has,
matching the matrix's "B (limited to health/watering)" row for roles that
can chat but only confirm a narrower set of actions -- enforced at the
tool-call layer by `AssistantToolRegistry`, not by gating the whole
confirm endpoint on a permission a read-only role would fail outright).

Ownership (a conversation belongs to the user who started it -- FR-9.4) is
enforced inside `AssistantConversationService`, not here (see that
service's own module docstring on why: a `NotFoundError` for
someone-else's conversation, never a 403 that would confirm the
conversation exists).
"""
from __future__ import annotations

import uuid
from typing import Any

from fastapi import APIRouter, Depends, Request, status

from app.api.deps import (
    PageParams,
    TenantContext,
    get_assistant_conversation_service,
    get_assistant_tool_registry,
    get_authorization_service,
    get_current_user,
    get_tenant_context,
    raise_if_denied,
    require_permission,
    request_context,
)
from app.ai.assistant.tool_registry import AssistantToolRegistry
from app.core.exceptions import ValidationError
from app.core.responses import ErrorResponse
from app.models.identity import User
from app.services.authorization_service import AuthorizationService
from app.schemas.ai import (
    AssistantConversationDetailResponse,
    AssistantConversationResponse,
    AssistantMessageResponse,
    ConfirmAssistantActionRequest,
    SendAssistantMessageRequest,
)
from app.services.assistant_conversation_service import AssistantConversationService

router = APIRouter()

_ERROR_RESPONSES: dict[int | str, dict[str, Any]] = {
    401: {"model": ErrorResponse, "description": "Authentication failed"},
    403: {"model": ErrorResponse, "description": "Missing ai_assistant:use / ai_assistant:confirm_write permission"},
    404: {"model": ErrorResponse, "description": "Conversation or message not found (or not owned by the caller)"},
}


def _require_org(tenant: TenantContext) -> uuid.UUID:
    if tenant.org_id is None:
        raise ValidationError("The user has no organization membership -- cannot use the AI Assistant.")
    return tenant.org_id


@router.post(
    "/ai/assistant/message", response_model=AssistantMessageResponse, status_code=status.HTTP_201_CREATED,
    responses={**_ERROR_RESPONSES, 503: {"model": ErrorResponse, "description": "The AI Assistant is temporarily unavailable"}},
    summary="Send a message to the AI Assistant (FR-9.1/9.2); omit conversation_id to start a new conversation",
)
async def send_assistant_message(
    body: SendAssistantMessageRequest, request: Request,
    user: User = Depends(get_current_user),
    tenant: TenantContext = Depends(get_tenant_context),
    conversation_service: AssistantConversationService = Depends(get_assistant_conversation_service),
    tool_registry: AssistantToolRegistry = Depends(get_assistant_tool_registry),
    _decision: Any = Depends(require_permission("ai_assistant:use")),
) -> AssistantMessageResponse:
    nursery_id = _require_org(tenant)
    message = await conversation_service.send_message(
        user=user, nursery_id=nursery_id, role_code=tenant.role_code, conversation_id=body.conversation_id,
        content=body.content, tool_registry=tool_registry, request_id=request_context(request).request_id,
    )
    return AssistantMessageResponse.model_validate(message)


@router.post(
    "/ai/assistant/actions/{message_id}/confirm", response_model=AssistantMessageResponse,
    responses={
        **_ERROR_RESPONSES,
        409: {"model": ErrorResponse, "description": "This action has already been confirmed or cancelled"},
        422: {"model": ErrorResponse, "description": "This message has no proposed action"},
    },
    summary="Confirm (or, with confirm=false, cancel) a proposed write action (FR-9.3)",
)
async def confirm_assistant_action(
    message_id: uuid.UUID, body: ConfirmAssistantActionRequest, request: Request,
    user: User = Depends(get_current_user),
    tenant: TenantContext = Depends(get_tenant_context),
    conversation_service: AssistantConversationService = Depends(get_assistant_conversation_service),
    tool_registry: AssistantToolRegistry = Depends(get_assistant_tool_registry),
    authz: AuthorizationService = Depends(get_authorization_service),
    _decision: Any = Depends(require_permission("ai_assistant:use")),
) -> AssistantMessageResponse:
    """
    `body.conversation_id` carries what the LLD's `{id}` path segment
    alone can't (the LLD names the confirm route `POST /ai/assistant/
    actions/{id}/confirm` without specifying whether `{id}` is the
    conversation or the message -- `{id}`=`message_id` here since that's
    what uniquely identifies "the one proposed action", with
    `conversation_id` carried in the body for the ownership check
    `AssistantConversationService._get_owned_conversation` performs).
    `body.confirm=false` cancels instead -- see `ConfirmAssistantAction
    Request`'s own docstring for why this is one endpoint, not two.
    """
    nursery_id = _require_org(tenant)
    if body.confirm:
        # `ai_assistant:confirm_write` is a STRICTER permission than `ai_assistant:use` (the matrix's "B
        # (limited to health/watering)" row) -- checked here in addition to the route-level `ai_assistant:use`
        # dependency above, not instead of it, so a read-only-chat role gets a clean permission error on
        # confirm rather than silently falling through to `AssistantToolRegistry.execute_confirmed_action`
        # (which re-checks the SPECIFIC tool's own permission, e.g. `watering:write` -- this check is the
        # coarser "can this user confirm ANYTHING at all" gate the matrix describes as a separate row).
        decision = await authz.authorize(
            user=user, permission="ai_assistant:confirm_write", resource_type="ai_assistant_message",
            resource_id=message_id, target_nursery_id=nursery_id, context=request_context(request),
        )
        if not decision.allowed:
            raise raise_if_denied(decision)
        result_message = await conversation_service.confirm_action(
            user=user, nursery_id=nursery_id, conversation_id=body.conversation_id, message_id=message_id,
            tool_registry=tool_registry, request_id=request_context(request).request_id,
        )
    else:
        result_message = await conversation_service.cancel_action(
            user=user, nursery_id=nursery_id, conversation_id=body.conversation_id, message_id=message_id,
            request_id=request_context(request).request_id,
        )
    return AssistantMessageResponse.model_validate(result_message)


@router.get(
    "/ai/assistant/conversations/{conversation_id}", response_model=AssistantConversationDetailResponse,
    responses=_ERROR_RESPONSES,
    summary="Get a conversation's metadata and paginated message history (FR-9.4)",
)
async def get_assistant_conversation(
    conversation_id: uuid.UUID, page_params: PageParams = Depends(),
    user: User = Depends(get_current_user), tenant: TenantContext = Depends(get_tenant_context),
    conversation_service: AssistantConversationService = Depends(get_assistant_conversation_service),
    _decision: Any = Depends(require_permission("ai_assistant:use")),
) -> AssistantConversationDetailResponse:
    nursery_id = _require_org(tenant)
    conversation, messages, total = await conversation_service.get_conversation(
        user=user, nursery_id=nursery_id, conversation_id=conversation_id,
        offset=page_params.offset, limit=page_params.page_size,
    )
    return AssistantConversationDetailResponse(
        conversation=AssistantConversationResponse.model_validate(conversation),
        messages=[AssistantMessageResponse.model_validate(m) for m in messages],
        total_messages=total,
    )

File: apps/web/lib/api/ai-assistant.ts
import { apiClient, unwrap } from "@/lib/api/client";
import type { components } from "@/lib/api/generated/schema";

/**
 * Thin typed wrappers for Module 10's AI Assistant surface
 * (`ai_assistant.py`) -- send a message (starts or continues a
 * conversation), confirm/cancel a proposed write action, and read one
 * conversation's full message history.
 *
 * There is no `GET /ai/assistant/conversations` list route -- a caller
 * can only ever fetch a conversation it already knows the id of (via
 * `get_assistant_conversation`'s own docstring: ownership is enforced as
 * a `NotFoundError`, not a 403, for someone-else's conversation). This is
 * why `store/assistant-store.ts` persists the current `conversationId`
 * client-side: it is the *only* way this frontend can reopen "my last
 * conversation" after a page reload, since the backend has nothing to
 * list from. If that id ever 404s (conversation deleted, or genuinely
 * belongs to a different account that happens to share this browser),
 * the panel starts a fresh conversation rather than erroring -- see
 * `components/assistant/assistant-panel.tsx`.
 */

export type AssistantMessageResponse = components["schemas"]["AssistantMessageResponse"];
export type AssistantConversationResponse = components["schemas"]["AssistantConversationResponse"];
export type AssistantConversationDetailResponse = components["schemas"]["AssistantConversationDetailResponse"];
export type SendAssistantMessageRequest = components["schemas"]["SendAssistantMessageRequest"];
export type ConfirmAssistantActionRequest = components["schemas"]["ConfirmAssistantActionRequest"];

/**
 * `AssistantMessageResponse.proposed_action` is `dict[str, Any] | null`
 * on the backend (no sub-schema) -- this hand-written interface mirrors
 * the real, exact shape `AssistantConversationService.send_message`
 * always constructs it with (`{tool_name, tool_arguments, summary}`, read
 * directly from that service's source, not guessed), cast via
 * `message.proposed_action as unknown as ProposedAction | null` in
 * `components/assistant/assistant-panel.tsx`. `action_status` is
 * likewise a free-text backend column (`String(30)`, not a DB enum) but
 * only ever set to one of these three literal values by that same
 * service, per its own inline comment.
 */
export interface ProposedAction {
  tool_name: string;
  tool_arguments: Record<string, unknown>;
  summary: string;
}
export type AssistantActionStatus = "pending_confirmation" | "confirmed" | "cancelled";

/** FR-9.1/9.2 -- omit `conversation_id` to start a new conversation; provide one you own to continue it. */
export async function sendAssistantMessage(body: SendAssistantMessageRequest): Promise<AssistantMessageResponse> {
  return unwrap(() => apiClient.POST("/api/v1/ai/assistant/message", { body }));
}

/** FR-9.3 -- `confirm: true` executes the proposed action through its normal service-layer validation; `confirm: false` discards it with no side effect. */
export async function confirmAssistantAction(
  messageId: string,
  body: ConfirmAssistantActionRequest,
): Promise<AssistantMessageResponse> {
  return unwrap(() =>
    apiClient.POST("/api/v1/ai/assistant/actions/{message_id}/confirm", { params: { path: { message_id: messageId } }, body }),
  );
}

/** FR-9.4 -- a conversation's metadata plus its paginated message history, newest-page-first per the real route's PageParams. */
export async function getAssistantConversation(
  conversationId: string,
  params: { page?: number; page_size?: number } = {},
): Promise<AssistantConversationDetailResponse> {
  return unwrap(() =>
    apiClient.GET("/api/v1/ai/assistant/conversations/{conversation_id}", {
      params: { path: { conversation_id: conversationId }, query: params },
    }),
  );
}
